File: client/src/protocol.ts
// Protocol types — mirrors server/internal/protocol/protocol.go
// Keep in sync manually.

export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades'
export type Rank = '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | 'J' | 'Q' | 'K' | 'A'

export interface Card {
  id: string
  rank: Rank
  suit: Suit
}

export interface CaptureOption {
  groups: Card[][]
}

export interface PublicPlayer {
  id: string
  name: string
  seatIndex: number
  avatarIndex: number // 1–6
  totalScore: number
  tablas: number
  capturedCount: number  // total cards captured this round
  scoringPoints: number  // card-point sum of captured cards this round
}

export type GamePhase = 'waiting' | 'playing' | 'round_end' | 'game_over'

export interface ClientGameState {
  roomId: string
  phase: GamePhase
  players: PublicPlayer[]
  tableCards: Card[]
  currentPlayerIndex: number
  lastCapturerIndex: number | null
  dealNumber: number
  roundNumber: number
  dealsRemaining: number // additional deals left in this round after the current one
  teamMode: boolean
}

export interface RoundScore {
  playerId: string
  cardPoints: number
  spilPoints: number
  tablaPoints: number
  total: number
}

// --- Client → Server ---

export interface CreateRoomMsg {
  type: 'CREATE_ROOM'
  playerName: string
  maxPlayers: 2 | 4
  avatarIndex: number // 1–6
}

export interface JoinRoomMsg {
  type: 'JOIN_ROOM'
  roomId: string
  playerName: string
  avatarIndex: number // 1–6
  reconnectToken?: string
}

export interface PlayCardMsg {
  type: 'PLAY_CARD'
  cardId: string
}

export interface ChooseCaptureMsg {
  type: 'CHOOSE_CAPTURE'
  optionIndex: number
}

export interface PingMsg {
  type: 'PING'
}

export type ClientMessage = CreateRoomMsg | JoinRoomMsg | PlayCardMsg | ChooseCaptureMsg | PingMsg

// --- Server → Client ---

export interface RoomJoinedMsg {
  type: 'ROOM_JOINED'
  roomId: string
  playerId: string
  reconnectToken: string
  seatIndex: number
}

export interface PlayerJoinedMsg {
  type: 'PLAYER_JOINED'
  player: PublicPlayer
}

export interface GameStartedMsg {
  type: 'GAME_STARTED'
  state: ClientGameState
}

export interface HandDealtMsg {
  type: 'HAND_DEALT'
  cards: Card[]
  dealsRemaining: number
}

export interface TurnStartMsg {
  type: 'TURN_START'
  playerIndex: number
}

export interface CardPlayedMsg {
  type: 'CARD_PLAYED'
  playerId: string
  card: Card
}

export interface CaptureOptionsMsg {
  type: 'CAPTURE_OPTIONS'
  options: CaptureOption[]
}

export interface CaptureMadeMsg {
  type: 'CAPTURE_MADE'
  playerId: string
  capturedCards: Card[]
  wasTabla: boolean
  capturedCount: number  // capturer's total this round after this capture
  scoringPoints: number  // capturer's scoring points this round after this capture
}

export interface CardDiscardedMsg {
  type: 'CARD_DISCARDED'
  card: Card
}

export interface RoundEndMsg {
  type: 'ROUND_END'
  scores: RoundScore[]
}

export interface GameOverMsg {
  type: 'GAME_OVER'
  winner: PublicPlayer
  players: PublicPlayer[]
}

export interface PlayerDisconnectedMsg {
  type: 'PLAYER_DISCONNECTED'
  playerId: string
}

export interface ErrorMsg {
  type: 'ERROR'
  code: string
  message: string
}

export interface PongMsg {
  type: 'PONG'
}

export interface TurnAutoSkippedMsg {
  type: 'TURN_AUTO_SKIPPED'
  playerId: string
  playerName: string
}

export type ServerMessage =
  | RoomJoinedMsg
  | PlayerJoinedMsg
  | GameStartedMsg
  | HandDealtMsg
  | TurnStartMsg
  | CardPlayedMsg
  | CaptureOptionsMsg
  | CaptureMadeMsg
  | CardDiscardedMsg
  | RoundEndMsg
  | GameOverMsg
  | PlayerDisconnectedMsg
  | ErrorMsg
  | PongMsg
  | TurnAutoSkippedMsg

File: server/internal/protocol/protocol.go
package protocol

import (
	"encoding/json"
	"fmt"

	"tablic/server/internal/game"
)

// --- Client → Server messages ---

type CreateRoomMsg struct {
	Type       string `json:"type"`
	PlayerName string `json:"playerName"`
	MaxPlayers int    `json:"maxPlayers"` // 2 or 4
}

type JoinRoomMsg struct {
	Type           string `json:"type"`
	RoomID         string `json:"roomId"`
	PlayerName     string `json:"playerName"`
	ReconnectToken string `json:"reconnectToken,omitempty"`
}

type PlayCardMsg struct {
	Type   string `json:"type"`
	CardID string `json:"cardId"`
}

type ChooseCaptureMsg struct {
	Type        string `json:"type"`
	OptionIndex int    `json:"optionIndex"`
}

type PingMsg struct {
	Type string `json:"type"`
}

// ParseClientMessage parses an incoming WebSocket message by its "type" field.
func ParseClientMessage(data []byte) (string, any, error) {
	var base struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(data, &base); err != nil {
		return "", nil, err
	}
	switch base.Type {
	case "CREATE_ROOM":
		var m CreateRoomMsg
		return base.Type, m, json.Unmarshal(data, &m)
	case "JOIN_ROOM":
		var m JoinRoomMsg
		return base.Type, m, json.Unmarshal(data, &m)
	case "PLAY_CARD":
		var m PlayCardMsg
		return base.Type, m, json.Unmarshal(data, &m)
	case "CHOOSE_CAPTURE":
		var m ChooseCaptureMsg
		return base.Type, m, json.Unmarshal(data, &m)
	case "PING":
		return base.Type, PingMsg{Type: "PING"}, nil
	default:
		return base.Type, nil, fmt.Errorf("unknown message type: %s", base.Type)
	}
}

// --- Server → Client messages ---

type RoomJoinedMsg struct {
	Type           string `json:"type"`
	RoomID         string `json:"roomId"`
	PlayerID       string `json:"playerId"`
	ReconnectToken string `json:"reconnectToken"`
	SeatIndex      int    `json:"seatIndex"`
}

type PlayerJoinedMsg struct {
	Type   string            `json:"type"`
	Player game.PublicPlayer `json:"player"`
}

type GameStartedMsg struct {
	Type  string               `json:"type"`
	State game.ClientGameState `json:"state"`
}

type HandDealtMsg struct {
	Type  string      `json:"type"`
	Cards []game.Card `json:"cards"`
}

type TurnStartMsg struct {
	Type        string `json:"type"`
	PlayerIndex int    `json:"playerIndex"`
}

type CardPlayedMsg struct {
	Type     string    `json:"type"`
	PlayerID string    `json:"playerId"`
	Card     game.Card `json:"card"`
}

type CaptureOptionsMsg struct {
	Type    string               `json:"type"`
	Options []game.CaptureOption `json:"options"`
}

type CaptureMadeMsg struct {
	Type          string      `json:"type"`
	PlayerID      string      `json:"playerId"`
	CapturedCards []game.Card `json:"capturedCards"`
	WasTabla      bool        `json:"wasTabla"`
}

type CardDiscardedMsg struct {
	Type string    `json:"type"`
	Card game.Card `json:"card"`
}

type RoundEndMsg struct {
	Type   string            `json:"type"`
	Scores []game.RoundScore `json:"scores"`
}

type GameOverMsg struct {
	Type    string              `json:"type"`
	Winner  game.PublicPlayer   `json:"winner"`
	Players []game.PublicPlayer `json:"players"`
}

type PlayerDisconnectedMsg struct {
	Type     string `json:"type"`
	PlayerID string `json:"playerId"`
}

type ErrorMsg struct {
	Type    string `json:"type"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

type PongMsg struct {
	Type string `json:"type"`
}

// MustMarshal marshals v to JSON; panics on error (only for server-controlled structs).
func MustMarshal(v any) []byte {
	b, err := json.Marshal(v)
	if err != nil {
		panic("protocol: failed to marshal: " + err.Error())
	}
	return b
}
